### DISCUSSION TITLE
Documentations for fission tool integration with other languages

### DISCUSSION THREAD

**Discussion by amnpandey** [fission/fission#3060] [Q&A] [answered]
Hello @soharab-ic
We have few question with respect to fission tool. Do we have any documentation on fission for developers. Basically, we would like to know:
1.  How to create fission resources using `GoLang` (if we are integrating it with go based component). How do we create fission resources like `environments`, `packages` and `functions`, triggers dynamically without using `fission cli`.
2. How to invoke the services. Is it through triggers?
3. Do we have fission library available for `golang`.
4. How do we build and create the `source`/`deployment` package if it is on github. Does fission support it?
5. Does the tool handle a high volume of concurrent executions without significant performance loss?
Could you please help us with answers to these. Thanks

**Comment by soharab-ic** [comment]
@amnpandey 

1. Please take a look this [Golang doc on Fission](https://fission.io/docs/usage/languages/go/).
2. Yes, you can invoke services through [triggers](https://fission.io/docs/usage/triggers/). Available triggers are: [HTTP trigger](https://fission.io/docs/usage/triggers/http-trigger/), [time trigger](https://fission.io/docs/usage/triggers/timer/) and [message queue based triggers](https://fission.io/docs/usage/triggers/message-queue-trigger-kind-keda/).
3. No available library for Golang. Please check how we build the runtime and builder environments for [Golang](https://github.com/fission/environments/tree/master/go). You can customize your own environment.
4. [url-as-archive-source](https://fission.io/docs/usage/function/url-as-archive-source/) when creating a `package` or a `function`. 
5. For high volume of concurrent executions, Please explore Fission executor types: [pool-manager](https://fission.io/blog/demystifying-fission-pool-manager/), [new-deploy](https://fission.io/blog/demystifying-fission-new-deploy/) and [container-as-a-function](https://fission.io/docs/usage/function/container-functions/).

You can explore [Fission blogs](https://fission.io/blog/) for more Fission's usecase in real-world.

**Comment by amnpandey** [comment]
@soharab-ic We have few additional questions which I have also posted in the fission public slack channel. Please help with these also. 
1. Do we have support for multi-tenancy in fission? Does it isolate functions in separate containers/VMs to prevent cross-access between tenants?
2. Can network policies (ingress and egress) be applied per tenant or at function level for network level security using any functionalities in fission?
3. We have two types of executors - PoolManager and New-Deployment. Which is recommended for high performance and running multiple functions in parallel?
4. Can different tenants have unique autoscaling rules based on their load requirements?
5. Is the tool compatible with CI/CD systems like Jenkins, GitLab, or GitHub Actions for automated deployment?
6. Can we build our own python runtime image and the builder image as mentioned in the language section: https://fission.io/docs/usage/languages/
7. Do we have any performance benchmarking for the tool. If so, can you please share us the metrics.


We want to run a python code with some pre-defined function name which accepts one input parameter of type dictionary and returns output of type json. Something like below:

```
def run(input: dict):
    return {"output_data": f'Hello,{input["name"]} !'}
```

Is this supported? How do we handle input parameters for the function and the output json type.

**Comment by sanketsudake** [comment, marked as answer]
> Do we have support for multi-tenancy in fission? Does it isolate functions in separate containers/VMs to prevent cross-access between tenants?

We can have multiple fission installations on the same cluster or use namespace-level isolation across a single fission installation.
To prevent cross-tenant communication, we can apply network policies so that cross-namespace communication is prevented.

> Can network policies (ingress and egress) be applied per tenant or at function level for network level security using any functionalities in fission?

Fission doesn't have any explicit support for the network policies but since Fission is Kubernetes native,
Fission would respect applied network policies as we run pods/deployments under the hood.

> We have two types of executors - PoolManager and New-Deployment. Which is recommended for high performance and running multiple functions in parallel?

Typically the pool manager strategy has more fine-grained control and is used by many fission users in production, where a single environment supports multiple functions.
Eg. a Python environment can have 100-150 functions running.
This supports an efficient usage of cluster resources. 
There are certain use cases where new deployment is also preferred. 
If you require a low cold-start time prefer pool manager. 

> Can different tenants have unique autoscaling rules based on their load requirements?

Autoscaling can be configured at the function/environment level. I am not sure if we have support at the fission global install level but something can be baked in here.

> Is the tool compatible with CI/CD systems like Jenkins, GitLab, or GitHub Actions for automated deployment?

Fission supports generating manifests/yaml spec which can be used to deploy functions. 
Users typically generate specs and apply them through APIs, Kubectl, or fission CLI. 
You can also generate a zip of the source and use it for deployment. 
So overall you can build a pipeline which can run on any CI/CD system.

> Can we build our own Python runtime image and the builder image as mentioned in the language section: https://fission.io/docs/usage/languages/

Yes, many fission users customize environments and use them for their purpose. Fission is extensible at all ends and many users to a lot of customizations in fission.
> 

You can find source for all environments here https://github.com/fission/environments

> Do we have any performance benchmarking for the tool? If so, can you please share the metrics?

We have done some benchmarking in the past https://github.com/fission/benchmark and we keep doing it from time to time. 
Depending on function execution time, we get different RPS stats. Some fission partners suppport benchmarking and finetuning suggestions as part of their enterprise offering.
https://fission.io/support/

**Comment by amnpandey** [comment]
Thanks @sanketsudake for your reply. Pls let us know on this as well as this is our main requirement.

>We want to run a python code with some pre-defined function name which accepts one input parameter of type dictionary and returns output of type json. Something like below:
def run(input: dict):
    return {"output_data": f'Hello,{input["name"]} !'}
Is this supported? How do we handle input parameters for the function and the output json type.

**Comment by sanketsudake** [comment]
So changes in the Python environment would be required accordingly.

[1] https://github.com/fission/environments/blob/master/python/server.py
[2] https://fission.io/blog/observability-with-opentelemetry-datadog-in-fission/#building-custom-python-environment

`/` endpoint is invoked on calling function https://github.com/fission/environments/blob/f26842188284bc69b0aa3276fcc4d3e587e7dd64/python/server.py#L174
and `/v2/specialize` is used to inject function code.
